### PR TITLE
Unset timestamp to fix cart error

### PR DESCRIPTION
**Pull request by Isengo1989**
### 1. Why is this change necessary?
Currently it is possible to buy a product, even if inactive, when it was in the cart beforehand. The validation is faulty due to an always set DataTimestamp

### 2. What does this change do, exactly?
It adds an unset -> set to **null** in case the lineItem is not in the productGateway anymore (therefor **inactive** or **deleted**)

### 3. Describe each step to reproduce the issue or behaviour.

1. Add a product to the cart
2. Set it inactive or delete it
3. Continue your shopping process and buy the product
4. Success

The cart should be validated accordingly and remove the lineItem from it. This is due to 

`$lineItem->setDataTimestamp(new \DateTimeImmutable());`

Being set when the item is initially being added, but never changed afterwards. The check

```
// product data was never detected and the product is not inside the data collection
        if ($product !== null || $item->getDataTimestamp() !== null) {
            return;
        }
```
will always lead to a return

### 4. Please link to the relevant issues (if any).
https://issues.shopware.com/issues/NEXT-17002

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/workflow/2020-08-03-implement-New-Changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

File: src/Core/Content/Product/Cart/ProductCartProcessor.php [hunk 1]
```diff
@@ -102,6 +102,8 @@ public function collect(CartDataCollection $data, Cart $original, SalesChannelCo
                     if (\in_array($lineItem->getReferencedId(), $products->getIds(), true)) {
                         $lineItem->setDataTimestamp(new \DateTimeImmutable());
                         $lineItem->setDataContextHash($hash);
+                    } else {
+                        $lineItem->setDataTimestamp(null);
```
Comment (shyim) [inline src/Core/Content/Product/Cart/ProductCartProcessor.php:106]: Can you add a test case for this?

Comment (Isengo1989) [reply src/Core/Content/Product/Cart/ProductCartProcessor.php:106]: Already on it

Testing Shopware\Core\Content\Test\Product\Cart\ProductCartProcessorTest
................................                                  33 / 33 (100%)

**Comment by Isengo1989** [comment]
@shyim I added 2 tests - one for the _delete_ case and one for the **inactive** case.

**Comment by J-Rahe** [comment]
Hey @Isengo1989, It seems that booth of your tests are failing against the current trunk. Could you please have a look at that?

**Comment by Isengo1989** [comment]
@J-Rahe as far as I see the test testProductFeaturesContainCorrectInformation is responsible for the failure:

removing the test leads to passing. Mine -> testInactiveProductCartRemoval and testDeletedProductCartRemoval do pass it.

```
PHPUnit 9.5.21 #StandWithUkraine

Random Seed:   1658500792

Testing Shopware\Core\Content\Test\Product\Cart\ProductCartProcessorTest
........................E.....S....                               35 / 35 (100%)

Time: 00:04.692, Memory: 118.50 MB

There was 1 error:

1) Shopware\Core\Content\Test\Product\Cart\ProductCartProcessorTest::testProductFeaturesContainCorrectInformation with data set #3 (array('customField', null, 'lorem_ipsum', 3), array(array(array('Default', array('Default')), array(array('Dolor sit amet.')))), array('lorem_ipsum', array('24c8b3e8cacc4bf2a743b8c5a7522a33', 'text', 'Dolor sit amet.'), 'customField'))
Doctrine\DBAL\Exception\InvalidFieldNameException: An exception occurred while executing 'SELECT `custom_field`.`id` as `custom_field.id`, `custom_field`.`name` as `custom_field.name`, `custom_field`.`type` as `custom_field.type`, `custom_field`.`config` as `custom_field.config`, `custom_field`.`active` as `custom_field.active`, `custom_field`.`set_id` as `custom_field.customFieldSetId`, `custom_field`.`allow_customer_write` as `custom_field.allowCustomerWrite`, `custom_field`.`created_at` as `custom_field.createdAt`, `custom_field`.`updated_at` as `custom_field.updatedAt` FROM `custom_field` WHERE (`custom_field`.`name` IN (?))' with params ["lorem_ipsum"]:

SQLSTATE[42S22]: Column not found: 1054 Unknown column 'custom_field.allow_customer_write' in 'field list'

/home/micha/platform_fork/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php:79
/home/micha/platform_fork/vendor/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php:182
/home/micha/platform_fork/vendor/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php:159
/home/micha/platform_fork/vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:2226
/home/micha/platform_fork/vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:1313
/home/micha/platform_fork/vendor/doctrine/dbal/lib/Doctrine/DBAL/Query/QueryBuilder.php:212
/home/micha/platform_fork/src/Core/Framework/DataAbstractionLayer/Dbal/EntityReader.php:299
/home/micha/platform_fork/src/Core/Framework/DataAbstractionLayer/Dbal/EntityReader.php:134
/home/micha/platform_fork/src/Core/Framework/DataAbstractionLayer/Dbal/EntityReader.php:86
/home/micha/platform_fork/src/Core/Content/Test/Category/Service/CountingEntityReader.php:35
/home/micha/platform_fork/src/Core/Framework/DataAbstractionLayer/EntityRepository.php:225
/home/micha/platform_fork/src/Core/Framework/DataAbstractionLayer/EntityRepository.php:254
/home/micha/platform_fork/src/Core/Framework/DataAbstractionLayer/EntityRepository.php:103
/home/micha/platform_fork/src/Core/Content/Product/Cart/ProductFeatureBuilder.php:141
/home/micha/platform_fork/src/Core/Content/Product/Cart/ProductFeatureBuilder.php:37
/home/micha/platform_fork/src/Core/Content/Product/Cart/ProductCartProcessor.php:125
/home/micha/platform_fork/src/Core/Profiling/Profiler.php:57
/home/micha/platform_fork/src/Core/Content/Product/Cart/ProductCartProcessor.php:126
/home/micha/platform_fork/src/Core/Checkout/Cart/Processor.php:101
/home/micha/platform_fork/src/Core/Checkout/Cart/Processor.php:65
/home/micha/platform_fork/src/Core/Profiling/Profiler.php:57
/home/micha/platform_fork/src/Core/Checkout/Cart/Processor.php:84
/home/micha/platform_fork/src/Core/Checkout/Cart/CartRuleLoader.php:127
/home/micha/platform_fork/src/Core/Profiling/Profiler.php:57
/home/micha/platform_fork/src/Core/Checkout/Cart/CartRuleLoader.php:177
/home/micha/platform_fork/src/Core/Checkout/Cart/CartRuleLoader.php:89
/home/micha/platform_fork/src/Core/Checkout/Cart/CartCalculator.php:30
/home/micha/platform_fork/src/Core/Profiling/Profiler.php:57
/home/micha/platform_fork/src/Core/Checkout/Cart/CartCalculator.php:39
/home/micha/platform_fork/src/Core/Checkout/Cart/SalesChannel/CartItemAddRoute.php:107
/home/micha/platform_fork/src/Core/Checkout/Cart/SalesChannel/CartService.php:122
/home/micha/platform_fork/src/Core/Content/Test/Product/Cart/ProductCartProcessorTest.php:767
/home/micha/platform_fork/src/Core/Content/Test/Product/Cart/ProductCartProcessorTest.php:322

Caused by
Doctrine\DBAL\Driver\PDO\Exception: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'custom_field.allow_customer_write' in 'field list'

/home/micha/platform_fork/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDO/Exception.php:18
/home/micha/platform_fork/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:119
/home/micha/platform_fork/vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:1304
/home/micha/platform_fork/vendor/doctrine/dbal/lib/Doctrine/DBAL/Query/QueryBuilder.php:212
/home/micha/platform_fork/src/Core/Framework/DataAbstractionLayer/Dbal/EntityReader.php:299
/home/micha/platform_fork/src/Core/Framework/DataAbstractionLayer/Dbal/EntityReader.php:134
/home/micha/platform_fork/src/Core/Framework/DataAbstractionLayer/Dbal/EntityReader.php:86
/home/micha/platform_fork/src/Core/Content/Test/Category/Service/CountingEntityReader.php:35
/home/micha/platform_fork/src/Core/Framework/DataAbstractionLayer/EntityRepository.php:225
/home/micha/platform_fork/src/Core/Framework/DataAbstractionLayer/EntityRepository.php:254
/home/micha/platform_fork/src/Core/Framework/DataAbstractionLayer/EntityRepository.php:103
/home/micha/platform_fork/src/Core/Content/Product/Cart/ProductFeatureBuilder.php:141
/home/micha/platform_fork/src/Core/Content/Product/Cart/ProductFeatureBuilder.php:37
/home/micha/platform_fork/src/Core/Content/Product/Cart/ProductCartProcessor.php:125
/home/micha/platform_fork/src/Core/Profiling/Profiler.php:57
/home/micha/platform_fork/src/Core/Content/Product/Cart/ProductCartProcessor.php:126
/home/micha/platform_fork/src/Core/Checkout/Cart/Processor.php:101
/home/micha/platform_fork/src/Core/Checkout/Cart/Processor.php:65
/home/micha/platform_fork/src/Core/Profiling/Profiler.php:57
/home/micha/platform_fork/src/Core/Checkout/Cart/Processor.php:84
/home/micha/platform_fork/src/Core/Checkout/Cart/CartRuleLoader.php:127
/home/micha/platform_fork/src/Core/Profiling/Profiler.php:57
/home/micha/platform_fork/src/Core/Checkout/Cart/CartRuleLoader.php:177
/home/micha/platform_fork/src/Core/Checkout/Cart/CartRuleLoader.php:89
/home/micha/platform_fork/src/Core/Checkout/Cart/CartCalculator.php:30
/home/micha/platform_fork/src/Core/Profiling/Profiler.php:57
/home/micha/platform_fork/src/Core/Checkout/Cart/CartCalculator.php:39
/home/micha/platform_fork/src/Core/Checkout/Cart/SalesChannel/CartItemAddRoute.php:107
/home/micha/platform_fork/src/Core/Checkout/Cart/SalesChannel/CartService.php:122
/home/micha/platform_fork/src/Core/Content/Test/Product/Cart/ProductCartProcessorTest.php:767
/home/micha/platform_fork/src/Core/Content/Test/Product/Cart/ProductCartProcessorTest.php:322

Caused by
PDOException: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'custom_field.allow_customer_write' in 'field list'

/home/micha/platform_fork/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:117
/home/micha/platform_fork/vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:1304
/home/micha/platform_fork/vendor/doctrine/dbal/lib/Doctrine/DBAL/Query/QueryBuilder.php:212
/home/micha/platform_fork/src/Core/Framework/DataAbstractionLayer/Dbal/EntityReader.php:299
/home/micha/platform_fork/src/Core/Framework/DataAbstractionLayer/Dbal/EntityReader.php:134
/home/micha/platform_fork/src/Core/Framework/DataAbstractionLayer/Dbal/EntityReader.php:86
/home/micha/platform_fork/src/Core/Content/Test/Category/Service/CountingEntityReader.php:35
/home/micha/platform_fork/src/Core/Framework/DataAbstractionLayer/EntityRepository.php:225
/home/micha/platform_fork/src/Core/Framework/DataAbstractionLayer/EntityRepository.php:254
/home/micha/platform_fork/src/Core/Framework/DataAbstractionLayer/EntityRepository.php:103
/home/micha/platform_fork/src/Core/Content/Product/Cart/ProductFeatureBuilder.php:141
/home/micha/platform_fork/src/Core/Content/Product/Cart/ProductFeatureBuilder.php:37
/home/micha/platform_fork/src/Core/Content/Product/Cart/ProductCartProcessor.php:125
/home/micha/platform_fork/src/Core/Profiling/Profiler.php:57
/home/micha/platform_fork/src/Core/Content/Product/Cart/ProductCartProcessor.php:126
/home/micha/platform_fork/src/Core/Checkout/Cart/Processor.php:101
/home/micha/platform_fork/src/Core/Checkout/Cart/Processor.php:65
/home/micha/platform_fork/src/Core/Profiling/Profiler.php:57
/home/micha/platform_fork/src/Core/Checkout/Cart/Processor.php:84
/home/micha/platform_fork/src/Core/Checkout/Cart/CartRuleLoader.php:127
/home/micha/platform_fork/src/Core/Profiling/Profiler.php:57
/home/micha/platform_fork/src/Core/Checkout/Cart/CartRuleLoader.php:177
/home/micha/platform_fork/src/Core/Checkout/Cart/CartRuleLoader.php:89
/home/micha/platform_fork/src/Core/Checkout/Cart/CartCalculator.php:30
/home/micha/platform_fork/src/Core/Profiling/Profiler.php:57
/home/micha/platform_fork/src/Core/Checkout/Cart/CartCalculator.php:39
/home/micha/platform_fork/src/Core/Checkout/Cart/SalesChannel/CartItemAddRoute.php:107
/home/micha/platform_fork/src/Core/Checkout/Cart/SalesChannel/CartService.php:122
/home/micha/platform_fork/src/Core/Content/Test/Product/Cart/ProductCartProcessorTest.php:767
/home/micha/platform_fork/src/Core/Content/Test/Product/Cart/ProductCartProcessorTest.php:322

ERRORS!
Tests: 35, Assertions: 173, Errors: 1, Skipped: 1.

```

**Comment by shyim** [comment]
run the unit tests again but with `FORCE_INSTALL=1` prepanded in the CLI

**Comment by Isengo1989** [comment]
@J-Rahe @shyim can you please link the error occuring. I just merged the trunk, dropped the DB & TEST_DB + added the FORCE_INSTALL, but still no error:


image

**Comment by J-Rahe** [comment]
```
There were 2 failures:
1) Shopware\Core\Content\Test\Product\Cart\ProductCartProcessorTest::testInactiveProductCartRemoval
Failed asserting that Shopware\Core\Checkout\Cart\LineItem\LineItem Object &000000007f3a49b70000000061abaa6f (
    'payload' => Array &0 (
        'isCloseout' => false
        'customFields' => Array &1 ()
        'createdAt' => '2022-06-30 08:56:08.847'
        'releaseDate' => null
        'isNew' => false
        'markAsTopseller' => null
        'purchasePrices' => '{"currencyId":"b7d2554b0ce847cd82f3ac9bd1c0dfca","net":5,"gross":7.5,"linked":false,"listPrice":null,"percentage":null,"regulationPrice":null,"extensions":[]}'
        'productNumber' => 'cf748fda4bea4475b39e4689c0357670'
        'manufacturerId' => null
        'taxId' => 'c8b6fc8ee4af45d0ab934843d329ce6c'
        'tagIds' => null
        'categoryIds' => null
        'propertyIds' => null
        'optionIds' => null
        'options' => Array &2 ()
        'streamIds' => null
        'parentId' => null
        'stock' => 10
        'features' => Array &3 ()
    )
    'id' => 'e1f5603f943e457f88266596f5a7ca5e'
    'referencedId' => 'e1f5603f943e457f88266596f5a7ca5e'
    'label' => 'test'
    'quantity' => 1
    'type' => 'product'
    'priceDefinition' => Shopware\Core\Checkout\Cart\Price\Struct\QuantityPriceDefinition Object &000000007f3dd3100000000061abaa6f (
        'price' => 15.0
        'taxRules' => Shopware\Core\Checkout\Cart\Tax\Struct\TaxRuleCollection Object &000000007f3d77440000000061abaa6f (
            'elements' => Array &4 (
                15 => Shopware\Core\Checkout\Cart\Tax\Struct\TaxRule Object &000000007f3a02910000000061abaa6f (
                    'taxRate' => 15.0
                    'percentage' => 100.0
                    'extensions' => Array &5 ()
                )
            )
            'extensions' => Array &6 ()
        )
        'quantity' => 1
        'isCalculated' => true
        'referencePriceDefinition' => null
        'listPrice' => null
        'regulationPrice' => null
        'extensions' => Array &7 ()
    )
    'price' => Shopware\Core\Checkout\Cart\Price\Struct\CalculatedPrice Object &000000007f3cbb030000000061abaa6f (
        'unitPrice' => 15.0
        'quantity' => 1
        'totalPrice' => 15.0
        'calculatedTaxes' => Shopware\Core\Checkout\Cart\Tax\Struct\CalculatedTaxCollection Object &000000007f3cbb060000000061abaa6f (
            'elements' => Array &8 (
                15 => Shopware\Core\Checkout\Cart\Tax\Struct\CalculatedTax Object &000000007f3ac17e0000000061abaa6f (
                    'tax' => 1.96
                    'taxRate' => 15.0
                    'price' => 15.0
                    'extensions' => Array &9 ()
                )
            )
            'extensions' => Array &10 ()
        )
        'taxRules' => Shopware\Core\Checkout\Cart\Tax\Struct\TaxRuleCollection Object &000000007f3d77440000000061abaa6f
        'referencePrice' => null
        'listPrice' => null
        'regulationPrice' => null
        'extensions' => Array &11 ()
    )
    'good' => true
    'description' => null
    'cover' => null
    'deliveryInformation' => Shopware\Core\Checkout\Cart\Delivery\Struct\DeliveryInformation Object &000000007f3a3b9d0000000061abaa6f (
        'stock' => 10
        'weight' => 100.0
        'freeDelivery' => false
        'restockTime' => null
        'deliveryTime' => null
        'height' => 101.0
        'width' => 102.0
        'length' => 103.0
        'extensions' => Array &12 ()
    )
    'children' => Shopware\Core\Checkout\Cart\LineItem\LineItemCollection Object &000000007f3d282f0000000061abaa6f (
        'elements' => Array &13 ()
        'extensions' => Array &14 ()
    )
    'requirement' => null
    'removable' => true
    'stackable' => true
    'quantityInformation' => Shopware\Core\Checkout\Cart\LineItem\QuantityInformation Object &000000007f3d7bae0000000061abaa6f (
        'minPurchase' => 1
        'maxPurchase' => 100
        'purchaseSteps' => 1
        'extensions' => Array &15 ()
    )
    'modified' => false
    'dataTimestamp' => DateTimeImmutable Object &000000007f3d77ad0000000061abaa6f (
        'date' => '2022-06-30 08:56:09.025988'
        'timezone_type' => 3
        'timezone' => 'UTC'
    )
    'dataContextHash' => 'bf9d265b33d183a6f75f93c4b04349a6'
    'extensions' => Array &16 ()
) is null.
/builds/shopware/6/product/platform/src/Core/Content/Test/Product/Cart/ProductCartProcessorTest.php:712
phpvfscomposer:///builds/shopware/6/product/platform/vendor/phpunit/phpunit/phpunit:97
2) Shopware\Core\Content\Test\Product\Cart\ProductCartProcessorTest::testDeletedProductCartRemoval
Failed asserting that Shopware\Core\Checkout\Cart\LineItem\LineItem Object &000000007f3aa6ce0000000061abaa6f (
    'payload' => Array &0 (
        'isCloseout' => false
        'customFields' => Array &1 ()
        'createdAt' => '2022-06-30 08:56:09.246'
        'releaseDate' => null
        'isNew' => false
        'markAsTopseller' => null
        'purchasePrices' => '{"currencyId":"b7d2554b0ce847cd82f3ac9bd1c0dfca","net":5,"gross":7.5,"linked":false,"listPrice":null,"percentage":null,"regulationPrice":null,"extensions":[]}'
        'productNumber' => '8d88f9cc9b5842b6be750f9a04e2d2a8'
        'manufacturerId' => null
        'taxId' => '8f9893d953384007bf37b37121607139'
        'tagIds' => null
        'categoryIds' => null
        'propertyIds' => null
        'optionIds' => null
        'options' => Array &2 ()
        'streamIds' => null
        'parentId' => null
        'stock' => 10
        'features' => Array &3 ()
    )
    'id' => '6a526f075314413eaff4d5f2701fecfe'
    'referencedId' => '6a526f075314413eaff4d5f2701fecfe'
    'label' => 'test'
    'quantity' => 1
    'type' => 'product'
    'priceDefinition' => Shopware\Core\Checkout\Cart\Price\Struct\QuantityPriceDefinition Object &000000007f3dac740000000061abaa6f (
        'price' => 15.0
        'taxRules' => Shopware\Core\Checkout\Cart\Tax\Struct\TaxRuleCollection Object &000000007f3a00560000000061abaa6f (
            'elements' => Array &4 (
                15 => Shopware\Core\Checkout\Cart\Tax\Struct\TaxRule Object &000000007f3cd4d70000000061abaa6f (
                    'taxRate' => 15.0
                    'percentage' => 100.0
                    'extensions' => Array &5 ()
                )
            )
            'extensions' => Array &6 ()
        )
        'quantity' => 1
        'isCalculated' => true
        'referencePriceDefinition' => null
        'listPrice' => null
        'regulationPrice' => null
        'extensions' => Array &7 ()
    )
    'price' => Shopware\Core\Checkout\Cart\Price\Struct\CalculatedPrice Object &000000007f3ae1730000000061abaa6f (
        'unitPrice' => 15.0
        'quantity' => 1
        'totalPrice' => 15.0
        'calculatedTaxes' => Shopware\Core\Checkout\Cart\Tax\Struct\CalculatedTaxCollection Object &000000007f3a93590000000061abaa6f (
            'elements' => Array &8 (
                15 => Shopware\Core\Checkout\Cart\Tax\Struct\CalculatedTax Object &000000007f3d7bac0000000061abaa6f (
                    'tax' => 1.96
                    'taxRate' => 15.0
                    'price' => 15.0
                    'extensions' => Array &9 ()
                )
            )
            'extensions' => Array &10 ()
        )
        'taxRules' => Shopware\Core\Checkout\Cart\Tax\Struct\TaxRuleCollection Object &000000007f3a00560000000061abaa6f
        'referencePrice' => null
        'listPrice' => null
        'regulationPrice' => null
        'extensions' => Array &11 ()
    )
    'good' => true
    'description' => null
    'cover' => null
    'deliveryInformation' => Shopware\Core\Checkout\Cart\Delivery\Struct\DeliveryInformation Object &000000007f3af4590000000061abaa6f (
        'stock' => 10
        'weight' => 100.0
        'freeDelivery' => false
        'restockTime' => null
        'deliveryTime' => null
        'height' => 101.0
        'width' => 102.0
        'length' => 103.0
        'extensions' => Array &12 ()
    )
    'children' => Shopware\Core\Checkout\Cart\LineItem\LineItemCollection Object &000000007f3aa1fc0000000061abaa6f (
        'elements' => Array &13 ()
        'extensions' => Array &14 ()
    )
    'requirement' => null
    'removable' => true
    'stackable' => true
    'quantityInformation' => Shopware\Core\Checkout\Cart\LineItem\QuantityInformation Object &000000007f3a86460000000061abaa6f (
        'minPurchase' => 1
        'maxPurchase' => 100
        'purchaseSteps' => 1
        'extensions' => Array &15 ()
    )
    'modified' => false
    'dataTimestamp' => DateTimeImmutable Object &000000007f3a90eb0000000061abaa6f (
        'date' => '2022-06-30 08:56:09.426624'
        'timezone_type' => 3
        'timezone' => 'UTC'
    )
    'dataContextHash' => 'bf9d265b33d183a6f75f93c4b04349a6'
    'extensions' => Array &16 ()
) is null.
```

**Comment by Isengo1989** [comment]
@J-Rahe were as in "past" ? I can not reproduce this error, can you please help me out here, we do have hundreds of those broken orders.

**Comment by keulinho** [comment]
Thx @Isengo1989 your changes were merged! :blue_heart:

**Pull request closed without merging**